presto-raptorx/src/main/java/com/facebook/presto/raptorx/storage/organization/OrganizationJob.java

```diff
@@ -43,66 +53,134 @@
+    public OrganizationJob(OrganizationSet organizationSet, ChunkCompactor compactor, MetadataWriter metadataWriter, TransactionManager transactionManager, long nodeId)
     {
         this.compactor = requireNonNull(compactor, "compactor is null");
         this.organizationSet = requireNonNull(organizationSet, "organizationSet is null");
-        this.raptorConnector = requireNonNull(raptorConnector, "transactionManager is null");
+        this.metadataWriter = requireNonNull(metadataWriter, "metadataWriter is null");
+        this.transactionManager = requireNonNull(transactionManager, "sequenceManager is null");
+        this.nodeId = nodeId;
     }
 
     @Override
     public void run()
     {
         try {
-            runJob(organizationSet.getTableId(), organizationSet.getBucketNumber(), organizationSet.getChunkIds());
+            runJob(organizationSet.getTableInfo(), organizationSet.getBucketNumber(), organizationSet.getChunkIds());
         }
         catch (IOException e) {
             throw new UncheckedIOException(e);
         }
     }
 
-    private void runJob(long tableId, int bucketNumber, Set<Long> chunkIds)
+    private void runJob(TableInfo tableInfo, int bucketNumber, Set<Long> chunkIds)
             throws IOException
     {
-        ConnectorTransactionHandle transactionHandle = raptorConnector.beginTransaction(IsolationLevel.REPEATABLE_READ, false);
-        Transaction transaction = raptorConnector.getTransactionManager().get(transactionHandle);
-        runJob(transaction, bucketNumber, tableId, chunkIds);
-        raptorConnector.commit(transactionHandle);
-    }
+        ConnectorTransactionHandle transactionHandle = transactionManager.create();
+        Transaction transaction = transactionManager.get(transactionHandle);
+        TableMetadata tableMetadata = getTableMetadata(tableInfo);
+        List<ChunkInfo> newChunks = performCompaction(transaction.getTransactionId(), bucketNumber, chunkIds, tableMetadata);
 
-    private void runJob(Transaction transaction, int bucketNumber, long tableId, Set<Long> oldChunkIds)
-            throws IOException
-    {
-        TableMetadata metadata = getTableMetadata(transaction, tableId);
-        List<ChunkInfo> newChunks = performCompaction(transaction, bucketNumber, oldChunkIds, metadata);
-        log.info("Compacted chunks %s into %s", oldChunkIds, newChunks.stream().map(ChunkInfo::getChunkId).collect(toList()));
-        transaction.insertChunks(tableId, newChunks);
-        transaction.deleteChunks(tableId, oldChunkIds);
+        long tableId = tableInfo.getTableId();
+        metadataWriter.insertWorkerTransaction(transaction.getTransactionId(), nodeId);
+        try {
+            metadataWriter.runWorkerTransaction(tableId, (dao) -> {
+                // 2 parties get this same LOCK. DatabaseMetadataWriter.writeShard(), and this.
+                Long lockedTableId = dao.getLockedTableId(tableId);
```

Comment:
It's hacky reusing an irrelevant table `table_sizes` to serve as an exclusive lock. 

Here we need to block organization when certain action is going on in the cluster (e.g. delete, drop table). Have you checked the `blockMaintenance()` logic in Raptor? AFAIK, it prevents compaction/organization from running when the table is under change.

Reply:
After discussion with Tao, I think maybe we should rename the `table_sizes` table to a more general name like `tables` or `table_stats`.

Also, we need to think about preventing background organization/compaction job failing users' DELETE or DROP TABLE queries. Ideally, organization/compaction should have lower priority than DELETE/DROP.

Reply:
compact vs DELETE INSERT is the most critical issue for now. I need some experiment to reach a solution. blockMaintenance may not be, based on early experiments. Actually V1 gets LOCK before blockMaintenance(), no clue it can prevent LOCK contention.

"tables" is already there. "table_sizes" is for sizes. We can use "table_stats", but not much differ. "tables" and  "table_sizes" both contain stats.

Reply:
Now we know that "blockMaintenance" is need in DELETE, but no need for INSERT.

For the LOCK, V2 is the same as V1, get a tableID row lock. Just V2 get in table_sizes, V1 get in tables